### PR TITLE
quic: populate ExtraValidationContext for cert validation

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: construct a TransportSocketCallbacks object for QUIC to populate ExtraValidationContext which can be used to access connection properties during cert validation.

Risk Level: low, ExtraValidationContext is not used in the default cert validator
Testing: added a new integration test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
@ggreenway PTAL.

File: source/common/quic/envoy_quic_client_session.cc [hunk 1]
```diff
@@ -13,15 +15,40 @@
 namespace Envoy {
 namespace Quic {
 
+// Quic connection doesn't use transport socket object to do handshake. This callback
+// implementation is used by the cert verifier to access Network::Connection interfaces.
+class ConnectionWrapper : public Network::TransportSocketCallbacks {
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_client_session.cc:20]: If this class is intended to be used for cert validation context, how about naming it something like "CertValidationContext"?

File: source/common/quic/envoy_quic_client_session.cc [hunk 2]
```diff
@@ -13,15 +15,40 @@
 namespace Envoy {
 namespace Quic {
 
+// Quic connection doesn't use transport socket object to do handshake. This callback
+// implementation is used by the cert verifier to access Network::Connection interfaces.
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_client_session.cc:19]: How about something like:

// This class is only used to provide extra context during certificate validation. As such
// it does not implement the various methods which are irrelevant to certificate validation.

**Review by RyanTheOptimist — COMMENTED** [review]
Oooh, nice! This is very clean.

**Comment by RyanTheOptimist** [comment]
Looks like coverage is failing.

**Review by ggreenway — DISMISSED** [review]
LGTM, modulo @RyanTheOptimist 's comments

File: source/common/quic/envoy_quic_client_session.cc (hunk 1 above)
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_session.cc]: done

File: source/common/quic/envoy_quic_client_session.cc [hunk 3]
```diff
@@ -15,11 +15,12 @@
 namespace Envoy {
 namespace Quic {
 
-// Quic connection doesn't use transport socket object to do handshake. This callback
-// implementation is used by the cert verifier to access Network::Connection interfaces.
-class ConnectionWrapper : public Network::TransportSocketCallbacks {
+// Quic connection doesn't use transport socket object to do handshake. This class is only used to
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_client_session.cc:18]: nit: Sorry, I meant to suggest removing this first sentence completely.

**Review by RyanTheOptimist — COMMENTED** [review]
LGTM! mod removing a sentence.

File: source/common/quic/envoy_quic_client_session.cc (hunk 3 above)
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_session.cc]: done

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
Looks like coverage is failing?
/wait

**Comment by danzh2010** [comment]
@RyanTheOptimist  mind taking another look?

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**